sgid_client/generators.py

```diff
@@ -0,0 +1,67 @@
+from typing import TypedDict
+from .error import Errors
+import secrets
+from base64 import urlsafe_b64encode
+import hashlib
+
+
+class GeneratePkcePairReturn(TypedDict):
+    code_verifier: str
+    code_challenge: str
+
+
+def generate_code_verifier(length=43) -> str:
+    """Generates the random code verifier
+
+    Args:
+        length (int, optional): The length of the code verifier to generate.
+        Defaults to 43.
+
+    Raises:
+        Exception: if length is <43 or >128
+
+    Returns:
+        str: The generated code verifier
+    """
+    if length < 43 or length > 128:
+        raise Exception(Errors["CODE_VERIFIER_LENGTH_ERROR"])
+    bytes = secrets.token_bytes(96)
+    encoded = urlsafe_b64encode(bytes)
+    return encoded.decode("utf-8")[:length]
+
+
+def generate_code_challenge(code_verifier: str) -> str:
+    """Calculates the S256 code challenge for a provided code verifier
+
+    Args:
+        code_verifier (str): The code verifier
+
+    Returns:
+        str: The calculated code challenge
+    """
+    verifier_bytearray = bytearray(code_verifier, encoding="utf-8")
```

Comment:
nit: might be misunderstanding something, but why aren't we using `ascii()` on the `code_verifier`? From the spec, the algorithm was:
```
code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
```

Reply:
ahaha good catch, no reason for this, it was just code I copy-pasted from somewhere to make it work, then used everywhere. updated everything to ascii in bc2f1d6